PYIC-7508: Check doc ID in auth source check

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed
Check doc ID in auth source check

### Why did it change

We want to be more sure that when using a previously attained driving licence VC to dodge the auth source check, that the document used is the same as that used at DCMAW.

This checks that the document identifiers in both VCs match, and if not route to an auth source check.

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-7508](https://govukverify.atlassian.net/browse/PYIC-7508)


[PYIC-7508]: https://govukverify.atlassian.net/browse/PYIC-7508?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-7508: Check doc ID in auth source check

We want to be more sure that when using a previously attained driving
licence VC to dodge the auth source check, that the document used is the
same as that used at DCMAW.

This checks that the document identifiers in both VCs match, and if not
route to an auth source check.

Joe-Edwards-GDS commented on lambdas/process-cri-callback/src/main/java/uk/gov/di/ipv/core/processcricallback/service/CriCheckingService.java:
```diff
@@ -259,29 +260,40 @@ public JourneyResponse checkVcResponse(
+                .filter(Objects::nonNull)
+                .filter(drivingPermitDetailsList -> !drivingPermitDetailsList.isEmpty())
+                .map(drivingPermitDetailsList -> drivingPermitDetailsList.get(0))
+                .map(DrivingPermitDetails::getPersonalNumber)
```
I wonder if we want to use the SPOT/CIMIT logic for this - https://govukverify.atlassian.net/wiki/spaces/Architecture/pages/3889922273/RFC+0042+Unique+identifiers+for+evidence+documents+for+de-duplicating+contra-indicators+in+Identity+Proofing+and+Verification+IPV#Driving-permits

That'll include the issuer etc., not just the number

(Up to you whether you want to implement the logic for _other_ document types)

Joe-Edwards-GDS commented on lambdas/process-cri-callback/src/main/java/uk/gov/di/ipv/core/processcricallback/service/CriCheckingService.java:
```diff
@@ -259,29 +260,40 @@ public JourneyResponse checkVcResponse(
+                .orElse(false);
+    }
+
+    private Optional<String> extractDrivingPermitIdentifier(
```
This is a bit of a strange compound method - I wonder if it would be clearer to split into
- `Optional<VC> findVcFromCri(List<VC> vcs, Cri cri)`
- `String getDrivingPermitIdentifier(VC vc)`

(not massively important...)

I do like splitting out `requiresAuthoritativeSourceCheck` though :)

Joe-Edwards-GDS commented on lambdas/process-cri-callback/src/main/java/uk/gov/di/ipv/core/processcricallback/service/CriCheckingService.java:
```diff
@@ -259,29 +260,40 @@ public JourneyResponse checkVcResponse(
-                .filter(vc -> Cri.DCMAW.equals(vc.getCri()))
-                .filter(
+                .filter(vc -> cri.equals(vc.getCri()))
+                .filter(successfulCheck ? VcHelper::isSuccessfulVc : x -> true)
```
Do we need this condition? I don't think we ever care about unsuccessful VCs...

I suppose you could argue that if you already had a _failed_ DL VC for the same doc id then we should treat that as a failure here, but that seems too complicated - might as well just send them on the DL auth source check.

Wynndow replied:
Is the problem here not that if they had a failed DL VC from earlier in the journey that matched identifiers, by not checking that the succuess state for that VC, the default is to _not_ send them for an auth source check?

Joe-Edwards-GDS replied:
Ah, I really meant the ternary as opposed to the success check itself - not that it matters now you've refactored :)

Wynndow committed:
PYIC-7508: Use full document identifier

This updates the logic to generate and compare the full document
identifier, as used by CIMIT and SPOT.

Joe-Edwards-GDS commented on lambdas/process-cri-callback/src/main/java/uk/gov/di/ipv/core/processcricallback/service/CriCheckingService.java:
```diff
@@ -259,29 +262,53 @@ public JourneyResponse checkVcResponse(
+                    permit.getIssuingCountry(),
+                    permit.getIssuedBy(),
+                    permit.getPersonalNumber(),
+                    permit.getIssueDate());
```
Do you need to default these to 'unknown' (or anything else) if they are missing?

Wynndow replied:
They'll default to a liternal null if they're not present, which I think is fine for our purposes here

Joe-Edwards-GDS commented on lambdas/process-cri-callback/src/main/java/uk/gov/di/ipv/core/processcricallback/service/CriCheckingService.java:
```diff
@@ -259,29 +262,53 @@ public JourneyResponse checkVcResponse(
+        LOGGER.warn(
+                LogHelper.buildLogMessage("Unable to get driving permit identifier from VC")
+                        .with(LOG_CRI_ID.getFieldName(), vc.getCri()));
+        return null;
```
This will mean that we'll skip the auth source check if this happens - but it's not going to work if the DL claim was missing anyway!

Happy to leave it as a warning - we could consider elevating to an error later if we want.

Wynndow replied:
Yep - I went back on fourth on this quite a bit about if we should be throwing in this scenario - and decided to just log. If we end up in there there's something pretty broken with the DL VC which will rear its head elsewhere.
